lib/HLSL/HLMatrixLowerPass.cpp

```diff
@@ -834,24 +834,32 @@ Value *HLMatrixLowerPass::lowerHLIntrinsic(CallInst *Call, IntrinsicOp Opcode) {
 
   // Delegate to a lowered intrinsic call
   SmallVector<Value*, 4> LoweredArgs;
+  SmallVector<Type*, 4> ArgTys;
   LoweredArgs.reserve(Call->getNumArgOperands());
+  ArgTys.reserve(Call->getNumArgOperands());
   for (Value *Arg : Call->arg_operands()) {
+    Value *LoweredArg = nullptr;
     if (Arg->getType()->isPointerTy()) {
       // ByRef parameter (for example, frexp's second parameter)
       // If the argument points to a lowered matrix variable, replace it here,
       // otherwise preserve the matrix type and let further passes handle the lowering.
-      Value *LoweredArg = tryGetLoweredPtrOperand(Arg, Builder);
+      LoweredArg = tryGetLoweredPtrOperand(Arg, Builder);
       if (LoweredArg == nullptr) LoweredArg = Arg;
-      LoweredArgs.emplace_back(LoweredArg);
     }
     else {
-      LoweredArgs.emplace_back(getLoweredByValOperand(Arg, Builder));
+      LoweredArg = getLoweredByValOperand(Arg, Builder);
     }
+    LoweredArgs.emplace_back(LoweredArg);
+    ArgTys.emplace_back(LoweredArg->getType());
   }
 
   Type *LoweredRetTy = HLMatrixType::getLoweredType(Call->getType());
-  return callHLFunction(*m_pModule, HLOpcodeGroup::HLIntrinsic, static_cast<unsigned>(Opcode), 
-    LoweredRetTy, LoweredArgs, Builder);
```

Comment:
Much of the new code in this area was lifted out of callHLFunction so I could access the call to GetOrCreateHLFunction and provide attributes.

Reply:
It could. It's called a lot of places which is why I backed away from that. Sometimes it's called to replace a previous function, in which case the attribs can be passed along. In other cases, it's a totally new function.

The only current use of these function attributes are for errors during codegen and the wave-sensitivity during the dxbreak cleanup. They might be useful for other things going forward though.